src/mbgl/renderer/layers/render_background_layer.cpp

```diff
@@ -152,4 +141,23 @@ optional<Color> RenderBackgroundLayer::getSolidBackground() const {
     return { evaluated.get<BackgroundColor>() * evaluated.get<BackgroundOpacity>() };
 }
 
+namespace {
+void addPatternIfNeeded(const std::string& id, const LayerPrepareParameters& params) {
+    if (!params.patternAtlas.getPattern(id)) {
+        if (auto* image = params.imageManager.getImage(id)) {
+            params.patternAtlas.addPattern(*image);
+        }
+    }  
+}
+} // namespace
+
+void RenderBackgroundLayer::prepare(const LayerPrepareParameters& params) {
+    const auto& evaluated = static_cast<const BackgroundLayerProperties&>(*evaluatedProperties).evaluated;
+    if (!evaluated.get<BackgroundPattern>().to.empty()) {
+        // Ensures that the texture gets added and uploaded to the atlas.
```

Comment:
How about:

// Ensures that the pattern bitmap gets copied to atlas bitmap. Atlas bitmap is uploaded to atlas texture in upload.

since addPatternIfNeeded is not uploading to texture.